src/Client.ts

```diff
@@ -1,350 +1,368 @@
+	 */
+	heartbeatInfo: AdvancedHeartbeatInfo = {
+		first: true,
+		acknowledged: false,
+		intervalTime: 0,
+		interval: null,
+	};
+
+	/**
+	 * Intents used by this client
+	 */
+	intents: Intents;
+
+	/**
+	 * Total number of members where the gateway will stop sending offline members in the guild member list
+	 */
+	largeThreshold: number;
+
+	/**
+	 * The rest manager of this client
+	 */
+	rest = new Rest(this);
+
+	/**
+	 * The last sequence number received from the WebSocket server
+	 */
+	seq: number | null = null;
+
+	/**
+	 * The session ID of this client
+	 */
+	sessionId?: string;
+
+	/**
+	 * The status of the connection of this client
+	 */
+	status = ClientStatus.Disconnected;
+
+	/**
+	 * The token used by this client
+	 */
+	token = process.env.DISCORD_CLIENT_TOKEN;
+
+	/**
+	 * The client user
+	 */
+	user?: APIUser;
+
+	/**
+	 * The user agent to append to requests to the API
+	 */
+	userAgent?: string;
+
+	/**
+	 * The websocket of this client
+	 */
+	ws?: WebSocket;
+
+	/**
+	 * @param options - Options for the client
+	 */
+	constructor({ intents, token, largeThreshold, userAgent }: ClientOptions) {
+		super();
+
+		this.intents = intents;
+		this.largeThreshold = largeThreshold ?? 50;
+		if (token != null) this.token = token;
+		this.userAgent = userAgent;
+	}
+
+	/**
+	 * Connect this client to the websocket.
+	 */
+	async connect(): Promise<void> {
+		if (this.status === ClientStatus.Disconnected) {
+			this.status = ClientStatus.Connecting;
+			this.ws = new WebSocket(await this.getGatewayUrl());
+			this.ws.on("message", (data: Buffer) => {
+				const payload = JSON.parse(data.toString()) as GatewayReceivePayload;
+
+				// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+				if (payload.s != null) this.seq = payload.s;
+
+				switch (payload.op) {
+					case GatewayOpcodes.Dispatch:
+						this._handleEvent(payload);
+						break;
+					case GatewayOpcodes.Heartbeat:
+						this._sendHeartbeat();
+						break;
+					case GatewayOpcodes.Reconnect:
+						this.ws!.close(5000, "Reconnecting");
+						void this.connect();
+						break;
+					case GatewayOpcodes.InvalidSession:
+						if (payload.d) this._resume();
+						if (this.status === ClientStatus.Resuming)
+							setTimeout(() => {
+								this._identify();
+							}, 1_000);
+						break;
+					case GatewayOpcodes.Hello:
+						this.heartbeatInfo.intervalTime = payload.d.heartbeat_interval;
+						this._sendHeartbeat();
+						this._heartbeat();
+						break;
+					case GatewayOpcodes.HeartbeatAck:
+						this.heartbeatInfo.acknowledged = true;
+						if (this.heartbeatInfo.first) {
+							this._identify();
+							this.heartbeatInfo.first = false;
+						}
+						break;
+					default:
+						this.emit(
+							"warn",
+							`Unknown opcode: ${(payload as { op: number }).op}`
+						);
+						break;
+				}
+			});
+		} else if (this.status === ClientStatus.Connected)
+			this.emit("error", new Error("Already connected"));

```

Comment:
If this doesn't throw the promise is fulfilled and the function is executed correctly while it shouldn't. We should throw in async functions when the problem is not related to an internal error (so not catchable)
```suggestion
			throw new Error("Already connected");
```

Reply:
In Node.JS, if an event is emitted with the Error class and the event is not listened to through the EventEmitter, it will throw an error anyway. So if someone does not want their connection to crash they can just listen to the 'error' event.